File: Commander/CommanderView.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using System.Windows.Forms;
using CefSharp;
using CefSharp.WinForms;

using Commander.Enums;
using Commander.Exceptions;
using Commander.Extension;
using Commander.Model;
using Commander.Processors;
using Commander.Properties;

namespace Commander
{
    class CommanderView
    {
        #region Properties

        public bool ShowHidden
        {
            get => _showHidden;
            set
            {
                _showHidden = value;
                Refresh();
            }
        }
        bool _showHidden;

        public string Path { get => currentItems.Path; }

        #endregion

        #region Types

        public enum ID
        {
            Left,
            Right
        }

        #endregion

        #region Constructor

        public CommanderView(ID id, IntPtr mainWindow, ChromiumWebBrowser browser, IHost host)
        {
            this.id = id;
            this.mainWindow = mainWindow;
            this.browser = browser;
            this.host = host;
        }

        #endregion

        #region Interface

        public async void Ready()
        {
            var path = host.RecentPath;
            var viewType = GetViewType(path);
            var columns = GetColumns(viewType);
            await ExecuteScriptAsync("setColumns", columns);
            ChangePath(path);
        }

        public async void ChangePath(string path, string directoryToSelect = null)
        {
            if (path == null)
                return;

            var request = requestFactory.Create();

            var viewType = GetViewType(path);
            var setColumns = viewType != currentItems.ViewType;

            var newItems = await Task.Factory.StartNew(() => 
            {
                switch (viewType)
                {
                    case ViewType.Root:
                        currentSorting = null;
                        return RootProcessor.Get();
                    default:
                        return DirectoryProcessor.Get(path, ShowHidden);
                }
            });

            if (request.IsCancelled)
                return;

            if (setColumns)
            {
                var columns = GetColumns(viewType);
                await ExecuteScriptAsync("setColumns", columns);
            }

            currentItems = newItems;

            Sort();

            host.RecentPath = currentItems.Path;

            int GetCurrentIndex()
            {
                if (directoryToSelect == null)
                    return ItemIndex.GetDefault(currentItems.ViewType);
                else if (viewType == ViewType.Directory)
                {
                    var folderToSelect = newItems.Directories.First(n => string.Compare(n.Name, directoryToSelect, true) == 0);
                    return ItemIndex.Create(ItemType.Directory, folderToSelect.Index);
                }
                else if (viewType == ViewType.Root)
                {
                    var folderToSelect = newItems.Drives.First(n => string.Compare(n.Name, directoryToSelect, true) == 0);
                    return ItemIndex.Create(ItemType.Directory, folderToSelect.Index);
                }
                else
                    return 0;
            }
            SetIndex(GetCurrentIndex());

            await ExecuteScriptWithParams("itemsChanged");

            if (currentItems.ViewType == ViewType.Directory)
            {
                var extended = await Task.Factory.StartNew(() => currentItems.ExtendItems());

                if (request.IsCancelled)
                    return;

                currentItems = extended;
                Sort();

                if (request.IsCancelled)
                    return;

                await ExecuteScriptWithParams("itemsChanged");
            }
        }

        public void ProcessItem(ProcessItemType type)
        {
            var itemType = ItemIndex.GetItemType(currentIndex);
            var item = GetCurrentItemPath(currentIndex);

            switch (itemType)
            {
                case ItemType.Parent:
                    var info = new DirectoryInfo(currentItems.Path);
                    ChangePath(item, info.Name);
                    break;
                case ItemType.Directory:
                    if (type == ProcessItemType.Properties)
                        ProcessFile(item, type);
                    else
                        ChangePath(item);
                    break;
                case ItemType.File:
                    ProcessFile(item, type);
                    break;
                default:
                    break;
            }
        }

        void ProcessFile(string file, ProcessItemType type)
        {
            switch (type)
            {
                case ProcessItemType.Show:
                    var p = new Process();
                    p.StartInfo.UseShellExecute = true;
                    p.StartInfo.ErrorDialog = true;
                    p.StartInfo.FileName = file;
                    p.Start();
                    break;
                case ProcessItemType.Properties:
                    var info = new ShellExecuteInfo()
                    {
                        lpVerb = "properties",
                        lpFile = file,
                        nShow = Api.SW_SHOW,
                        fMask = Api.SEE_MASK_INVOKEIDLIST
                    };
                    info.cbSize = Marshal.SizeOf(info);
                    Api.ShellExecuteEx(ref info);
                    break;
                case ProcessItemType.StartAs:
                    p = Process.Start("rundll32.exe", $"shell32, OpenAs_RunDLL {file}");
                    break;
            }
        }

        public string GetItems()
        {
            var sw = new Stopwatch();
            sw.Start();

            IEnumerable<ResponseItem> resultItems;
            switch (currentItems.ViewType)
            {
                case ViewType.Root:
                    resultItems = currentItems.Drives.Select(n => 
                        new ResponseItem(ItemType.Directory, ItemIndex.Create(ItemType.Directory, n.Index), 
                        new[] {
                            n.Name,
                            n.Label,
                            n.Size.ToString("N0")
                        }, "Drive", currentIndex.IsSelected(n.Index, ItemType.Directory)));
                    break;
                default:
                    resultItems = DirectoryProcessor.GetItems(currentItems, currentIndex);
                    break;
            }

            var response = new Response(currentItems.Path, resultItems);
            var result = Json.Serialize(response);
            var elapsed = sw.Elapsed;
            Debugger.Log(1, "Main", $"JSON conversion duration: {elapsed}\n");
            return result;
        }

        public async void SetIndex(int index)
        {
            currentIndex = index;
            await ExecuteScriptAsync("setCurrentItem", GetCurrentItemPath(currentIndex));
        }

        public async void Sort(int index, bool ascending)
        {
            currentSorting = (index, !ascending);
            Sort();
            await ExecuteScriptWithParams("itemsChanged");
        }

        public async void CreateFolder(string item)
        {
            try
            {
                var viewType = GetViewType(Path);
                switch (viewType)
                {
                    case ViewType.Directory:
                        DirectoryProcessor.CreateFolder(Path, item);
                        break;
                    default:
                        break;
                }
                ChangePath(Path, item);
            }
            catch (AlreadyExistsException)
            {
                await browser.EvaluateScriptAsync($"commander.showDialog", Resources.FolderAlreadyExists);
            }
        }

        public async void Copy(string targetPath)
        {
            var fileop = new SHFILEOPSTRUCT()
            {
                fFlags = FILEOP_FLAGS.FOF_NOCONFIRMATION | FILEOP_FLAGS.FOF_NOCONFIRMMKDIR | FILEOP_FLAGS.FOF_MULTIDESTFILES,
                hwnd = mainWindow,
                lpszProgressTitle = "Commander",
                wFunc = FileFuncFlags.FO_COPY,
            };
        }

        public void SetSelected(object[] objects)
            => selectedIndexes = objects?.Cast<int>().ToArray();

        #endregion

        #region Methods

        public void Refresh() => ChangePath(currentItems.Path);

        // TODO: call only in the right folder
        // TODO: Dialog resources (Abbrechen) 
        public async Task CreateFolder()
            //TODO: dialog.inputText = item.items[0] != ".." ? item.items[0] : ""
            => await ExecuteScriptWithParams("createFolder", Resources.dialogCreateFolder);

        public async Task Copy(CommanderView otherView)
            => await ExecuteScriptWithParams("copy", otherView.Path, Resources.dialogCopy);

        Columns GetColumns(ViewType viewType)
        {
            switch (viewType)
            {
                case ViewType.Root:
                    return new Columns(RootProcessor.Name, new[]
                    {
                        new Column(Resources.RootName, false),
                        new Column(Resources.RootLabel, false),
                        new Column(Resources.RootSize, false, ColumnsType.Size)
                    });
                default:
                    return new Columns(DirectoryProcessor.Name, new[]
                    {
                        new Column(Resources.DirectoryName),
                        new Column(Resources.DirectoryExtension),
                        new Column(Resources.DirectoryDate, true, ColumnsType.Date),
                        new Column(Resources.DirectorySize, true, ColumnsType.Size),
                        new Column(Resources.DirectoryVersion)
                    });
            }
        }

        async Task ExecuteScriptAsync(string method, object param)
        {
            var sw = new Stopwatch();
            sw.Start();

            var json = Json.Serialize(param);
            var elapsed = sw.Elapsed;
            await browser.EvaluateScriptAsync($"{host.Class}.{method}({json})");

            var elapsed2 = sw.Elapsed;
            Debugger.Log(1, "Main", $"Script execution duration: {elapsed}, {elapsed2}\n");
        }

        string GetCurrentItemPath(int index)
        {
            var itemType = ItemIndex.GetItemType(index);
            var arrayIndex = ItemIndex.GetArrayIndex(index);

            string GetDirectoryItemPath()
            {
                switch (itemType)
                {
                    case ItemType.Directory:
                        return currentItems.Directories[arrayIndex].Name;
                    case ItemType.File:
                        return currentItems.Files[arrayIndex].Name + currentItems.Files[arrayIndex].Extension;
                    case ItemType.Parent:
                        var info = new DirectoryInfo(currentItems.Path);
                        return info.Parent?.FullName ?? "root";
                    default:
                        return null;
                }
            }

            string GetDirectory()
            {
                var directory = GetDirectoryItemPath();
                return directory == "root" ? "root" : System.IO.Path.Combine(currentItems.Path, directory);
            }

            return currentItems.ViewType == ViewType.Root 
                ? currentItems.Drives[arrayIndex].Name 
                : GetDirectory();
        }

        async Task ExecuteScriptWithParams(string method, params object[] parameters)
        {
            var sw = new Stopwatch();
            sw.Start();

            await browser.EvaluateScriptAsync($"{host.Class}.{method}", parameters);

            var elapsed = sw.Elapsed;
            Debugger.Log(1, "Main", $"Script execution duration: {elapsed}\n");
        }

        ViewType GetViewType(string path)
        {
            if (path == RootProcessor.Name || (path.EndsWith("..") && path.Length == 5))
                return ViewType.Root;
            else
                return ViewType.Directory;
        }

        void Sort()
        {
            if (currentSorting.HasValue && currentItems.Files != null)
            {
                IEnumerable<FileItem> sortedItems;
                switch (currentSorting.Value.index)
                {
                    case 0:
                        sortedItems = currentItems.Files.OrderBy(n => n.Name);
                        break;
                    case 1:
                        sortedItems = currentItems.Files.OrderBy(n => n.Extension).ThenBy(n => n.Name);
                        break;
                    case 2:
                        sortedItems = currentItems.Files.OrderBy(n => n.Date).ThenBy(n => n.Name);
                        break;
                    case 3:
                        sortedItems = currentItems.Files.OrderBy(n => n.Size).ThenBy(n => n.Name);
                        break;
                    case 4:
                        sortedItems = currentItems.Files.OrderBy(n => n.Version, FileVersion.Comparer).ThenBy(n => n.Name);
                        break;
                    default:
                        sortedItems = currentItems.Files;
                        break;
                }
                if (currentSorting.Value.descending)
                    sortedItems = sortedItems.Reverse();

                currentItems = Items.UpdateFiles(currentItems, sortedItems);
            }
        }

        #endregion

        #region Fields

        readonly ID id;
        readonly ChromiumWebBrowser browser;
        readonly IHost host;
        readonly RequestFactory requestFactory = new RequestFactory();
        readonly IntPtr mainWindow;
        
        Items currentItems;
        int currentIndex;

        (int index, bool descending)? currentSorting = null;
        int[] selectedIndexes;

        #endregion
    }
}

File: Commander/CommanderView.fs
namespace Commander

open System
open System.IO
open System.Windows.Forms

open CefSharp
open Model
open DirectoryProcessor
open System.Diagnostics
open ClrWinApi
open System.Runtime.InteropServices

[<NoComparison>]
[<NoEquality>]
type BrowserAccess = {
    GetRecentPath: unit->string
    SetRecentPath: string->unit
    ExecuteScript: string->obj option->Async<JavascriptResponse>
    ExecuteCommanderScript: string->obj option->Async<JavascriptResponse>
    ExecuteScriptWithParams: string->obj[]->Async<JavascriptResponse>
    MainWindow: nativeint
    Dispatcher: Control
}

type CommanderView(browserAccess: BrowserAccess) as this =
    let requestFactory = RequestFactory()
    let mutable currentItems = Model.createEmptyItems ()
    let mutable currentIndex = 0
    let mutable currentSorting: (int*bool) option = None
    let mutable selectedIndexes: int[] = [||]
    
    let getViewType (path: string) = 
        match path with 
        | RootProcessor.name -> ViewType.Root
        | _ when path.EndsWith("..") && path.Length = 5 -> ViewType.Root
        | _ -> ViewType.Directory

    let getColumns viewType (directoryToSelect: string option) = 
        match viewType with
        | ViewType.Root -> { Name = RootProcessor.name; Values = RootProcessor.columns }
        | ViewType.Directory | _ -> { Name = DirectoryProcessor.name; Values = DirectoryProcessor.columns }

    let getCurrentItemPath index = 
        let itemType = ItemIndex.getItemType index
        let arrayIndex = ItemIndex.getArrayIndex index

        let getDirectoryItemPath () = 
            match itemType with
            | ItemType.Directory -> currentItems.Directories.[arrayIndex].Name
            | ItemType.File -> currentItems.Files.[arrayIndex].Name + currentItems.Files.[arrayIndex].Extension
            | ItemType.Parent ->
                let info = new DirectoryInfo(currentItems.Path)
                if info.Parent <> null then 
                    info.Parent.FullName 
                else 
                    "root"
            | _ -> null
    
        let getDirectory () =
            let directory = getDirectoryItemPath ()
            if directory = "root" then 
                "root" 
            else System.IO.Path.Combine(currentItems.Path, directory)

        if currentItems.ViewType = ViewType.Root then
            currentItems.Drives.[arrayIndex].Name 
        else
            getDirectory ()

    let processFile (file: string) processItemType = 
        match processItemType with
        | ProcessItemType.Show ->  
            use p = new Process()
            p.StartInfo.UseShellExecute <- true
            p.StartInfo.ErrorDialog <- true
            p.StartInfo.FileName <- file
            p.Start () |> ignore
        | ProcessItemType.Properties ->  
            let mutable info = ShellExecuteInfo()
            info.Verb <- "properties"
            info.File <- file
            info.Show <- ShowWindowFlag.Show
            info.Mask <- ShellExecuteFlag.InvokeIDList
            info.Size <- Marshal.SizeOf(info)
            ShellExecuteEx(info) |> ignore
        | ProcessItemType.StartAs ->  
            Process.Start("rundll32.exe", "shell32, OpenAs_RunDLL "+ file) |> ignore
        | _ -> ()

    let changePath path (directoryToSelect: string option) = 
        let request = requestFactory.create()
        let viewType = getViewType path
        let setColumns = viewType <> currentItems.ViewType 
        let get = 
            match viewType with
            | ViewType.Root ->
                currentSorting <- None
                RootProcessor.get
            | ViewType.Directory | _ ->
                DirectoryProcessor.get path Globals.showHidden
        async {
            let newItems = get ()

            if setColumns && not request.IsCancelled then
                let columns = getColumns viewType None
                let! response = browserAccess.ExecuteScript "setColumns" (Some (columns:> obj))
                ()
            if not request.IsCancelled then  
                currentItems <- newItems
                browserAccess.SetRecentPath currentItems.Path

                let getCurrentIndex () =
                    match directoryToSelect, viewType with
                    | None, ViewType.Root -> ItemIndex.create ItemType.Drive 0
                    | None, _ -> ItemIndex.getDefault currentItems.ViewType
                    | Some value, ViewType.Directory -> 
                        let folderToSelect = 
                            newItems.Directories
                            |> Seq.tryFind (fun n -> String.Compare(n.Name, value, true) = 0)
                        match folderToSelect with
                        | Some value -> ItemIndex.create ItemType.Directory value.Index
                        | None -> 0
                    | Some value, ViewType.Root -> 
                        let folderToSelect = newItems.Drives |> Array.find (fun n -> String.Compare(n.Name, value, true) = 0)
                        folderToSelect.Index
                    | _, _ -> 0

                let! res = this.SetIndex (getCurrentIndex ())

                let! response = browserAccess.ExecuteScript "itemsChanged" None
                if viewType = ViewType.Directory then
                    async {
                        let newItems =
                            if not request.IsCancelled then  
                                Some (extendItems currentItems)
                            else 
                                None
                        match request.IsCancelled, newItems with
                        | false, Some value -> 
                            currentItems <- sortBy currentSorting value
                            let! response = browserAccess.ExecuteScript "itemsChanged" None
                            ()
                        | _ -> ()
                        
                    } |> Async.Start
        } |> Async.Start
        ()

    let sort (index: int) (ascending: bool) =
        currentSorting <- Some (index, not ascending)
        currentItems <- sortBy currentSorting currentItems
        browserAccess.ExecuteScript "itemsChanged" None

    let getSelectedItems () = 
        let selectedIndexes = 
            match selectedIndexes.Length with
            | 0 -> [| currentIndex |]
            | _ -> selectedIndexes

        selectedIndexes |> Seq.map (fun n -> (ItemIndex.getArrayIndex n, ItemIndex.getItemType n))

    member this.Ready () = 
        changePath (browserAccess.GetRecentPath ()) None 

    member this.ProcessItem processItemType = 
        let itemType = ItemIndex.getItemType currentIndex
        let item = getCurrentItemPath currentIndex

        match itemType, processItemType with
            | ItemType.Parent, _ -> 
                let info = new DirectoryInfo(currentItems.Path)
                changePath item (Some info.Name)
            | ItemType.Directory, ProcessItemType.Properties -> processFile item processItemType
            | ItemType.Directory, _ -> changePath item None
            | ItemType.File, _ -> processFile item processItemType
            | ItemType.Drive, _ -> changePath item None
            | _ -> ()
        ()

    member val internal Other = this with get, set

    member this.GetItems () = 
        let responses = 
            match currentItems.ViewType, currentItems.Drives, currentItems.Directories, currentItems.Files with
            | ViewType.Root, drives, [||], [||] -> 
                drives
                |> Seq.mapi (fun i n -> createDriveResponse n.Name n.Label n.Size n.Index (ItemIndex.isSelected currentIndex i ItemType.Drive))
                |> Seq.toArray
            | ViewType.Directory, [||], directories, files -> getItems currentIndex directories files
            | _ -> failwith "Invalid ViewType"
        let response: Response = { 
            Path = currentItems.Path
            Items = responses 
        }
        Json.serialize response

    member this.SetIndex index =
        currentIndex <- index
        browserAccess.ExecuteScript "setCurrentItem" (Some ((getCurrentItemPath currentIndex):>obj))

    member this.ChangePath path = changePath path None

    member this.Sort index ascending = sort index ascending

    member this.StartCopy (otherView: CommanderView) = 
        browserAccess.ExecuteScriptWithParams "copy" [| otherView.Path :> obj; Resources.Resources.dialogCopy :> obj |] |> ignore

    member this.StartCreateFolder () = 
        //TODO: dialog.inputText = item.items[0] != ".." ? item.items[0] : ""
        browserAccess.ExecuteScript "createFolder" (Some (Resources.Resources.dialogCreateFolder :> obj)) |> ignore

    member this.AdaptPath (path: string) = ()
    member this.Path with get() = currentItems.Path
    member this.Refresh () = changePath currentItems.Path None
    member this.SetSelected (selectedValues: obj[]) = 
        let toInt (n: obj) = 
            match n with
            | null -> 0
            | _ -> n :?> int
        match selectedValues with 
        | null -> selectedIndexes <- [||]
        | _ -> selectedIndexes <-  selectedValues |> Array.map toInt
        
    member this.CreateFolder (item: string) =
        try 
            match currentItems.ViewType with
            | ViewType.Directory -> 
                DirectoryProcessor.createFolder currentItems.Path item
                changePath currentItems.Path (Some item)
            | _ -> ()
        with Exceptions.AlreadyExists ->
            browserAccess.ExecuteCommanderScript "showDialog" (Some (Resources.Resources.FolderAlreadyExists :> obj)) |> ignore

    member this.Copy (targetPath: string) = 
        async { 
            let selectedItems = getSelectedItems ()
            let index, itemType = selectedItems |> Seq.head
            let! refresh = 
                match itemType with
                | ItemType.Directory | ItemType.File -> 
                    DirectoryProcessor.copy currentItems selectedItems targetPath browserAccess.MainWindow browserAccess.Dispatcher
                | _ -> async.Return false
            if refresh then this.Other.Refresh() 
        } |> Async.Start

    member this.GetTestItems() = DirectoryProcessor.getTestItems ()